src/librustc/middle/infer/error_reporting.rs

```diff
@@ -491,6 +496,49 @@ impl<'a, 'tcx> ErrorReporting<'tcx> for InferCtxt<'a, 'tcx> {
         }
     }
 
+    /// Adds a note if the types come from similarly named crates
+    fn check_and_note_conflicting_crates(&self, terr: &ty::TypeError<'tcx>, sp: Span) {
+        let report_path_match = |did1: DefId, did2: DefId| {
+            // Only external crates, if either is from a local
+            // module we could have false positives
+            if !(did1.is_local() || did2.is_local()) {
```

Comment:
You should check that the crates are different - for example, `fn main() { let _ = [{struct Foo; Foo}, {struct Foo; Foo}]; }` will have the path "main::Foo" for both of the two different types.